_posts/2017/2017-16-01-testing-async-interactions.md

```diff
@@ -0,0 +1,246 @@
+We also reduce runtime dependencies between our services.
+
+For the sample application, we use [Spring Boot](http://projects.spring.io/spring-boot/),  [RabbitMQ](https://www.rabbitmq.com/) as the message broker, and [Spring Amqp](http://projects.spring.io/spring-amqp/) to interact with it from our application.
+
+The code referenced in this post is available on [github](https://github.com/mduesterhoeft/testing-asynchonous-interactions). It is advisable to make yourself familiar with this code before reading on.
+
+Whenever we create a [product](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/product-service/src/main/java/com/epages/product/Product.java) an event is [published](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/product-service/src/main/java/com/epages/product/ProductEventHandler.java) which the checkout service [picks up](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/main/java/com/epages/checkout/ProductSubscriber.java).
+It will then store [its view on the product](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/main/java/com/epages/checkout/ProductRef.java) in the checkout database.
+
+## A primitive test is better than nothing
+
+How should we test this interaction between the product and the checkout service?
+Of course, an integration test would be possible.
+But we favor fast feedback and ideally want issues to become apparent as early as possible and without having to start all our services.
+So we strive for a solution that allows us to test the interactions on the checkout service in isolation.
+We do not want to run the product service and the message broker to be able to test it.
+We call such a test a local integration test.
+
+A test that is fairly easy to implement on checkout-service side is to just hard-code the event payload as the checkout service expects it and assert that the service correctly consumes this event.
+
+Such a [test](https://github.com/mduesterhoeft/testing-asynchonous-interactions/blob/master/checkout-service/src/test/java/com/epages/checkout/ProductSubscriberHardcodedPayloadTest.java) should do two things:
+
+- make sure that the JSON event payload can be deserialized correctly
+- assert that the service is reacting appropriately
+
+```java
+@Test
+public void should_handle_product_created_event() {
+  //GIVEN
+  String payload = "{\n" +
+    "  \"id\": 8,\n" +
+    "  \"name\": \"Awesome Jeans\",\n" +
+    "  \"salesPrice\": 49.99\n" +
+    "}";
+
+  //WHEN
+  createAndEmitEvent(payload);
+  //THEN
+  then(productRefRepository.exists(8L)).isTrue();
+}
+```
+
+Note that the test uses a mock connection factory to avoid connecting to a running message broker.
+It also uses the Spring Amqp `MessageListener` to emit the event. Thus we leave the deserialization to Spring Amqp and thus also test the configuration there.
```

Comment:
Thus we leave the deserialization to Spring Amqp and also test the configuration there.